support/ebpf/interpreter_dispatcher.ebpf.c

```diff
@@ -124,6 +126,75 @@ bpf_map_def SEC("maps") apm_int_procs = {
   .max_entries = 128,
 };
 
+bpf_map_def SEC("maps") go_procs = {
+  .type        = BPF_MAP_TYPE_HASH,
+  .key_size    = sizeof(pid_t),
+  .value_size  = sizeof(GoCustomLabelsOffsets),
+  .max_entries = 128,
+};
+
+static inline __attribute__((__always_inline__)) void *
+get_m_ptr(struct GoCustomLabelsOffsets *offs, UnwindState *state)
+{
+  long res;
+
+  size_t g_addr;
+#if defined(__x86_64__)
+  u64 g_addr_offset = 0xfffffffffffffff8;
+  void *tls_base    = NULL;
+  res               = tsd_get_base(&tls_base);
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed to get tsd base; can't read m_ptr");
+    return NULL;
+  }
+
+  res = bpf_probe_read_user(&g_addr, sizeof(void *), (void *)((u64)tls_base + g_addr_offset));
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed to read g_addr, tls_base(%lx)", (unsigned long)tls_base);
+    return NULL;
+  }
+#elif defined(__aarch64__)
+  g_addr = state->r28;
+#endif
+
+  DEBUG_PRINT("cl: reading m_ptr_addr at 0x%lx + 0x%x", g_addr, offs->m_offset);
+  void *m_ptr_addr;
+  res = bpf_probe_read_user(&m_ptr_addr, sizeof(void *), (void *)(g_addr + offs->m_offset));
+  if (res < 0) {
+    DEBUG_PRINT("cl: failed m_ptr_addr");
+    return NULL;
+  }
+
+  return m_ptr_addr;
+}
+
+static inline __attribute__((__always_inline__)) void
+maybe_add_go_custom_labels(struct pt_regs *ctx, PerCPURecord *record)
+{
+  u32 pid = record->trace.pid;
+  // The Go label extraction code is too big to fit in this program, so we need to
+  // tail call it, in order to keep the hashing and clearing code in this program it
+  // will tail call back to us with this bool set.
+  if (!record->state.processed_go_labels) {
+    GoCustomLabelsOffsets *offsets = bpf_map_lookup_elem(&go_procs, &pid);
+    if (!offsets) {
+      DEBUG_PRINT("cl: no offsets, %d not recognized as a go binary", pid);
+      return;
+    }
+
+    void *m_ptr_addr = get_m_ptr(offsets, &record->state);
+    if (!m_ptr_addr) {
+      return;
+    }
+    record->customLabelsState.go_m_ptr = m_ptr_addr;
+
+    DEBUG_PRINT("cl: trace is within a process with Go custom labels enabled");
+    increment_metric(metricID_UnwindGoCustomLabelsAttempts);
+    record->state.processed_go_labels = true;
+    tail_call(ctx, PROG_GO_LABELS);
```

Comment:
Could we instead hook the main entry point to do a lookup from `go_labels` and store the `PROG_GO_LABELS` as the stop program ID if needed. Would avoid an extra tail call from the curren STOP -> GO_LABELS -> STOP round trip.